src/laser_simulator.cpp

```diff
@@ -65,25 +66,27 @@ void LaserScannerSimulator::update_loop(const ros::TimerEvent& event)
     get_laser_pose(&l_x,&l_y,&l_theta);
     //ROS_INFO_STREAM_THROTTLE(2,"x: " << l_x << " y: " << l_y << " theta: " <<  l_theta);
     update_scan(l_x,l_y,l_theta);
-    output_scan.header.stamp = event.current_real;
-    laser_pub.publish(output_scan);
+    output_scan.header.stamp = clock_->now();
+    laser_pub->publish(output_scan);
 }
 
 void LaserScannerSimulator::get_map()
 {
-    nav_msgs::GetMapRequest req;
-    nav_msgs::GetMapResponse resp;
-    if (ros::service::call(map_service, req, resp))
-    {
-        map = resp.map;
-        ROS_INFO_STREAM("Got a " << map.info.width << "x" << map.info.height << " map with resolution " << map.info.resolution);
-        have_map = true;
-    }
-    else 
-    {
-        ROS_WARN_THROTTLE(10,"No map received - service '/static_map' not available (will publish only max_range)");
-        have_map = false;
-    }
+    RCLCPP_INFO(logger_, "Requesting the map");
+    auto req = std::make_shared<nav_msgs::srv::GetMap::Request>();
+    nav_msgs::srv::GetMap::Response resp;
+    auto cb = [this](rclcpp::Client<nav_msgs::srv::GetMap>::SharedFuture future){
+      assert(future.valid());
+      auto result = future.get();
+      map = result->map;
+      RCLCPP_INFO_STREAM(logger_, "Got a " << map.info.width << "x" << map.info.height << " map with resolution " << map.info.resolution);
+      have_map = true;
+    };
+    auto future = map_service_->async_send_request(req, cb);
+
+    // TODO(Ramon): How to log a service error?
```

Comment:
I don't know how to get a callback when the service failed. If I fix this then the node is fully ported

Reply:
You can't? https://github.com/ros2/ros2/issues/921

Reply:
Strange

I added the link to that issue.